spec/spec_helper.cr

```diff
@@ -33,40 +33,33 @@ end
 
 record SemanticResult,
   program : Program,
-  node : ASTNode,
-  type : Type
+  node : ASTNode
 
-def assert_type(str, flags = nil, inject_primitives = true)
-  result = semantic_result(str, flags, inject_primitives: inject_primitives)
+def assert_type(str, *, inject_primitives = true, flags = nil, file = __FILE__, line = __LINE__)
+  result = semantic(str, flags: flags, inject_primitives: inject_primitives)
   program = result.program
   expected_type = with program yield program
-  result.type.should eq(expected_type)
+  node = result.node
+  if node.is_a?(Expressions)
+    node = node.last
+  end
```

Comment:
Is this needed? Expression's type should be already the type of the last expression. It's fine to leave it like this, though, if there's some other reason I'm not seeing.

Reply:
Yes, I also thought this shouldn't be necessary and the return type should be expected to be the last one.
But without that, this spec fails:

https://github.com/crystal-lang/crystal/blob/499d08d471aab7ac95b4ba3fa8187fe932658756/spec/compiler/semantic/no_return_spec.cr#L21

Maybe that's actually a bug in the type resolution for `Expressions` that it doesn't take `NoReturn` into account?
I didn't look into it. This same behaviour was in `semantic_result` before this refactor, so I just copied it over.

https://github.com/crystal-lang/crystal/blob/499d08d471aab7ac95b4ba3fa8187fe932658756/spec/spec_helper.cr#L67

Reply:
Good point. Maybe `CleanupTransformer` should check that if an Expressions has no type, give it `NoReturn` type. We do that in other places.

But don't worry about it, it can be part of a separate PR or not at all.

Reply:
Definitely.
FTR: It has a type, it's just the type of the last expression, even if there's a `NoReturn`:
```
     Failure/Error: assert_type("lib LibC; fun exit : NoReturn; end; LibC.exit; 1") { no_return }

       Expected: NoReturn
            got: Int32
```